clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -450,7 +451,21 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
     private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, long correlationId,
                                long now) {
         Errors error = response.error;
-        if (error != Errors.NONE) {
+        if (error == Errors.MESSAGE_TOO_LARGE && batch.recordCount > 1) {
+            // If the batch is too large, we split the batch and send the split batches again. We do not decrement
+            // the retry attempts in this case.
+            log.warn("Got RecordBatchTooLargeException in correlation id {} on topic-partition {}, spitting and retrying ({} attempts left). Error: {}",
+                     correlationId,
+                     batch.topicPartition,
+                     this.retries - batch.attempts(),
+                     error);
+            // Reset the estimated compression ratio to the initial value. There are several different ways to do the
+            // reset. We chose the most conservative one to ensure the split doesn't happen too often.
+            CompressionRatioEstimator.resetEstimation(batch.topicPartition.topic());
+            this.accumulator.splitAndReenqueue(batch);
+            this.accumulator.deallocate(batch);
```

Comment:
I may have missed this, but for the split batches I don't see any special handling for _not_ deallocating them. i.e., since those were created off-the-book, shouldn't their deallocation not be accounted in the accumulator? The off-accumulator batches should not be returned to the accumulator.